src/3d/symbols/qgspolygon3dsymbol_p.cpp

```diff
@@ -262,28 +260,81 @@ static Qt3DRender::QCullFace::CullingMode _qt3DcullingMode( Qgs3DTypes::CullingM
+Qt3DRender::QMaterial *QgsPolygon3DSymbolHandler::material( const QgsPolygon3DSymbol &symbol, bool isSelected, const Qgs3DRenderContext &context ) const
 {
-  Qt3DExtras::QPhongMaterial *material = new Qt3DExtras::QPhongMaterial;
+  Qt3DRender::QMaterial *retMaterial = nullptr;
+  if ( symbol.material().isUsingDiffuseTexture() )
+  {
+    QString textureFilePath = symbol.material().texturePath();
+    Qt3DExtras::QDiffuseMapMaterial *material = new Qt3DExtras::QDiffuseMapMaterial;
+
+    // front/back side culling
+    auto techniques = material->effect()->techniques();
+    for ( auto tit = techniques.constBegin(); tit != techniques.constEnd(); ++tit )
+    {
+      auto renderPasses = ( *tit )->renderPasses();
+      for ( auto rpit = renderPasses.begin(); rpit != renderPasses.end(); ++rpit )
+      {
+        Qt3DRender::QCullFace *cullFace = new Qt3DRender::QCullFace;
+        cullFace->setMode( _qt3DcullingMode( symbol.cullingMode() ) );
+        ( *rpit )->addRenderState( cullFace );
+      }
+    }
+    Qt3DRender::QTextureImage *textureImage = new Qt3DRender::QTextureImage;
+    textureImage->setSource( QUrl::fromLocalFile( textureFilePath ) );
+    material->diffuse()->addTextureImage( textureImage );
+
+    material->diffuse()->wrapMode()->setX( Qt3DRender::QTextureWrapMode::Repeat );
+    material->diffuse()->wrapMode()->setY( Qt3DRender::QTextureWrapMode::Repeat );
+    material->diffuse()->wrapMode()->setZ( Qt3DRender::QTextureWrapMode::Repeat );
+    material->setSpecular( symbol.material().specular() );
+    material->setShininess( symbol.material().shininess() );
+    material->setTextureScale( symbol.material().textureScale() );
+
+    if ( isSelected )
+    {
+      // update the material with selection colors
+      // TODO : dampen the color of diffuse texture
+//      mat->setDiffuse( context.map().selectionColor() );
+      material->setAmbient( context.map().selectionColor().darker() );
+    }
 
-  // front/back side culling
-  auto techniques = material->effect()->techniques();
-  for ( auto tit = techniques.constBegin(); tit != techniques.constEnd(); ++tit )
+    retMaterial = material;
+
+  }
+  else
   {
-    auto renderPasses = ( *tit )->renderPasses();
-    for ( auto rpit = renderPasses.begin(); rpit != renderPasses.end(); ++rpit )
+
+    Qt3DExtras::QPhongMaterial *material = new Qt3DExtras::QPhongMaterial;
+
+    // front/back side culling
+    auto techniques = material->effect()->techniques();
+    for ( auto tit = techniques.constBegin(); tit != techniques.constEnd(); ++tit )
+    {
+      auto renderPasses = ( *tit )->renderPasses();
+      for ( auto rpit = renderPasses.begin(); rpit != renderPasses.end(); ++rpit )
+      {
+        Qt3DRender::QCullFace *cullFace = new Qt3DRender::QCullFace;
+        cullFace->setMode( _qt3DcullingMode( symbol.cullingMode() ) );
+        ( *rpit )->addRenderState( cullFace );
+      }
+    }
```

Comment:
can we move this code to set culling mode to a separate function - that would avoid having two copies of it...